File: backend/app/schemas/financial.py
"""Financial data Pydantic schemas."""

from datetime import datetime, date
from typing import Optional, List, Dict, Any
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field, validator

from app.schemas.base import BaseSchema, TimestampSchema, UUIDSchema, PaginatedResponse


# Financial Report Schemas
class FinancialReportBase(BaseModel):
    """Base financial report schema."""
    
    ticker: str = Field(..., max_length=10, description="Stock ticker")
    fiscal_year: int = Field(..., ge=1900, le=2100, description="Fiscal year")
    fiscal_period: str = Field(..., description="Fiscal period")
    report_type: str = Field(..., description="Report type")
    announced_at: datetime = Field(..., description="Announcement date")
    source_url: Optional[str] = Field(None, max_length=500, description="Source URL")
    
    @validator('fiscal_period')
    def validate_fiscal_period(cls, v):
        """Validate fiscal period."""
        valid_periods = ['Q1', 'Q2', 'Q3', 'Q4', 'FY', 'H1', 'H2']
        if v not in valid_periods:
            raise ValueError(f'Fiscal period must be one of: {", ".join(valid_periods)}')
        return v
    
    @validator('report_type')
    def validate_report_type(cls, v):
        """Validate report type."""
        valid_types = ['earnings', 'balance_sheet', 'cash_flow', 'income_statement']
        if v not in valid_types:
            raise ValueError(f'Report type must be one of: {", ".join(valid_types)}')
        return v


class FinancialReportCreate(FinancialReportBase):
    """Financial report creation schema."""
    
    data: Dict[str, Any] = Field(..., description="Financial data")


class FinancialReport(BaseSchema, FinancialReportBase, TimestampSchema):
    """Financial report response schema."""
    
    data: Dict[str, Any] = Field(..., description="Financial data")


# Earnings Schemas
class EarningsBase(BaseModel):
    """Base earnings schema."""
    
    ticker: str = Field(..., max_length=10, description="Stock ticker")
    fiscal_year: int = Field(..., ge=1900, le=2100, description="Fiscal year")
    fiscal_quarter: int = Field(..., ge=1, le=4, description="Fiscal quarter")
    earnings_date: date = Field(..., description="Earnings announcement date")
    revenue: Optional[Decimal] = Field(None, description="Revenue")
    net_income: Optional[Decimal] = Field(None, description="Net income")
    eps_actual: Optional[Decimal] = Field(None, description="Actual EPS")
    eps_estimate: Optional[Decimal] = Field(None, description="EPS estimate")
    eps_surprise: Optional[Decimal] = Field(None, description="EPS surprise")
    revenue_estimate: Optional[Decimal] = Field(None, description="Revenue estimate")
    revenue_surprise: Optional[Decimal] = Field(None, description="Revenue surprise")


class EarningsCreate(EarningsBase):
    """Earnings creation schema."""
    pass


class Earnings(BaseSchema, EarningsBase, TimestampSchema):
    """Earnings response schema."""
    pass


# Balance Sheet Schemas
class BalanceSheetBase(BaseModel):
    """Base balance sheet schema."""
    
    ticker: str = Field(..., max_length=10, description="Stock ticker")
    fiscal_year: int = Field(..., ge=1900, le=2100, description="Fiscal year")
    fiscal_period: str = Field(..., description="Fiscal period")
    report_date: date = Field(..., description="Report date")
    
    # Assets
    total_assets: Optional[Decimal] = Field(None, description="Total assets")
    current_assets: Optional[Decimal] = Field(None, description="Current assets")
    cash_and_equivalents: Optional[Decimal] = Field(None, description="Cash and cash equivalents")
    accounts_receivable: Optional[Decimal] = Field(None, description="Accounts receivable")
    inventory: Optional[Decimal] = Field(None, description="Inventory")
    property_plant_equipment: Optional[Decimal] = Field(None, description="Property, plant & equipment")
    
    # Liabilities
    total_liabilities: Optional[Decimal] = Field(None, description="Total liabilities")
    current_liabilities: Optional[Decimal] = Field(None, description="Current liabilities")
    accounts_payable: Optional[Decimal] = Field(None, description="Accounts payable")
    short_term_debt: Optional[Decimal] = Field(None, description="Short-term debt")
    long_term_debt: Optional[Decimal] = Field(None, description="Long-term debt")
    
    # Equity
    total_equity: Optional[Decimal] = Field(None, description="Total shareholders' equity")
    retained_earnings: Optional[Decimal] = Field(None, description="Retained earnings")
    
    @validator('fiscal_period')
    def validate_fiscal_period(cls, v):
        """Validate fiscal period."""
        valid_periods = ['Q1', 'Q2', 'Q3', 'Q4', 'FY']
        if v not in valid_periods:
            raise ValueError(f'Fiscal period must be one of: {", ".join(valid_periods)}')
        return v


class BalanceSheetCreate(BalanceSheetBase):
    """Balance sheet creation schema."""
    pass


class BalanceSheet(BaseSchema, BalanceSheetBase, TimestampSchema):
    """Balance sheet response schema."""
    pass


# Income Statement Schemas
class IncomeStatementBase(BaseModel):
    """Base income statement schema."""
    
    ticker: str = Field(..., max_length=10, description="Stock ticker")
    fiscal_year: int = Field(..., ge=1900, le=2100, description="Fiscal year")
    fiscal_period: str = Field(..., description="Fiscal period")
    report_date: date = Field(..., description="Report date")
    
    # Revenue
    total_revenue: Optional[Decimal] = Field(None, description="Total revenue")
    cost_of_revenue: Optional[Decimal] = Field(None, description="Cost of revenue")
    gross_profit: Optional[Decimal] = Field(None, description="Gross profit")
    
    # Operating
    operating_expenses: Optional[Decimal] = Field(None, description="Operating expenses")
    operating_income: Optional[Decimal] = Field(None, description="Operating income")
    
    # Other
    interest_expense: Optional[Decimal] = Field(None, description="Interest expense")
    interest_income: Optional[Decimal] = Field(None, description="Interest income")
    other_income: Optional[Decimal] = Field(None, description="Other income")
    
    # Income
    income_before_tax: Optional[Decimal] = Field(None, description="Income before tax")
    income_tax_expense: Optional[Decimal] = Field(None, description="Income tax expense")
    net_income: Optional[Decimal] = Field(None, description="Net income")
    
    # Per Share
    basic_eps: Optional[Decimal] = Field(None, description="Basic EPS")
    diluted_eps: Optional[Decimal] = Field(None, description="Diluted EPS")
    weighted_average_shares: Optional[int] = Field(None, description="Weighted average shares")
    
    @validator('fiscal_period')
    def validate_fiscal_period(cls, v):
        """Validate fiscal period."""
        valid_periods = ['Q1', 'Q2', 'Q3', 'Q4', 'FY']
        if v not in valid_periods:
            raise ValueError(f'Fiscal period must be one of: {", ".join(valid_periods)}')
        return v


class IncomeStatementCreate(IncomeStatementBase):
    """Income statement creation schema."""
    pass


class IncomeStatement(BaseSchema, IncomeStatementBase, TimestampSchema):
    """Income statement response schema."""
    pass


# Cash Flow Schemas
class CashFlowBase(BaseModel):
    """Base cash flow schema."""
    
    ticker: str = Field(..., max_length=10, description="Stock ticker")
    fiscal_year: int = Field(..., ge=1900, le=2100, description="Fiscal year")
    fiscal_period: str = Field(..., description="Fiscal period")
    report_date: date = Field(..., description="Report date")
    
    # Operating Activities
    net_income: Optional[Decimal] = Field(None, description="Net income")
    depreciation_amortization: Optional[Decimal] = Field(None, description="Depreciation & amortization")
    changes_working_capital: Optional[Decimal] = Field(None, description="Changes in working capital")
    operating_cash_flow: Optional[Decimal] = Field(None, description="Operating cash flow")
    
    # Investing Activities
    capital_expenditures: Optional[Decimal] = Field(None, description="Capital expenditures")
    investments: Optional[Decimal] = Field(None, description="Investments")
    investing_cash_flow: Optional[Decimal] = Field(None, description="Investing cash flow")
    
    # Financing Activities
    debt_issuance: Optional[Decimal] = Field(None, description="Debt issuance")
    debt_repayment: Optional[Decimal] = Field(None, description="Debt repayment")
    dividends_paid: Optional[Decimal] = Field(None, description="Dividends paid")
    share_repurchases: Optional[Decimal] = Field(None, description="Share repurchases")
    financing_cash_flow: Optional[Decimal] = Field(None, description="Financing cash flow")
    
    # Net Change
    net_change_cash: Optional[Decimal] = Field(None, description="Net change in cash")
    free_cash_flow: Optional[Decimal] = Field(None, description="Free cash flow")
    
    @validator('fiscal_period')
    def validate_fiscal_period(cls, v):
        """Validate fiscal period."""
        valid_periods = ['Q1', 'Q2', 'Q3', 'Q4', 'FY']
        if v not in valid_periods:
            raise ValueError(f'Fiscal period must be one of: {", ".join(valid_periods)}')
        return v


class CashFlowCreate(CashFlowBase):
    """Cash flow creation schema."""
    pass


class CashFlow(BaseSchema, CashFlowBase, TimestampSchema):
    """Cash flow response schema."""
    pass


# Financial Ratios Schemas
class FinancialRatiosBase(BaseModel):
    """Base financial ratios schema."""
    
    ticker: str = Field(..., max_length=10, description="Stock ticker")
    fiscal_year: int = Field(..., ge=1900, le=2100, description="Fiscal year")
    fiscal_period: str = Field(..., description="Fiscal period")
    calculation_date: date = Field(..., description="Calculation date")
    
    # Profitability Ratios
    gross_margin: Optional[float] = Field(None, description="Gross margin")
    operating_margin: Optional[float] = Field(None, description="Operating margin")
    net_margin: Optional[float] = Field(None, description="Net margin")
    return_on_assets: Optional[float] = Field(None, description="Return on assets")
    return_on_equity: Optional[float] = Field(None, description="Return on equity")
    
    # Liquidity Ratios
    current_ratio: Optional[float] = Field(None, description="Current ratio")
    quick_ratio: Optional[float] = Field(None, description="Quick ratio")
    cash_ratio: Optional[float] = Field(None, description="Cash ratio")
    
    # Leverage Ratios
    debt_to_equity: Optional[float] = Field(None, description="Debt-to-equity ratio")
    debt_to_assets: Optional[float] = Field(None, description="Debt-to-assets ratio")
    interest_coverage: Optional[float] = Field(None, description="Interest coverage ratio")
    
    # Efficiency Ratios
    asset_turnover: Optional[float] = Field(None, description="Asset turnover")
    inventory_turnover: Optional[float] = Field(None, description="Inventory turnover")
    receivables_turnover: Optional[float] = Field(None, description="Receivables turnover")
    
    @validator('fiscal_period')
    def validate_fiscal_period(cls, v):
        """Validate fiscal period."""
        valid_periods = ['Q1', 'Q2', 'Q3', 'Q4', 'FY']
        if v not in valid_periods:
            raise ValueError(f'Fiscal period must be one of: {", ".join(valid_periods)}')
        return v


class FinancialRatiosCreate(FinancialRatiosBase):
    """Financial ratios creation schema."""
    pass


class FinancialRatios(BaseSchema, FinancialRatiosBase, TimestampSchema):
    """Financial ratios response schema."""
    pass


# Financial Summary Schemas
class FinancialSummary(BaseModel):
    """Financial summary schema."""
    
    ticker: str = Field(..., description="Stock ticker")
    company_name: str = Field(..., description="Company name")
    latest_fiscal_year: int = Field(..., description="Latest fiscal year")
    latest_fiscal_period: str = Field(..., description="Latest fiscal period")
    
    # Latest Financial Data
    latest_revenue: Optional[Decimal] = Field(None, description="Latest revenue")
    latest_net_income: Optional[Decimal] = Field(None, description="Latest net income")
    latest_eps: Optional[Decimal] = Field(None, description="Latest EPS")
    latest_total_assets: Optional[Decimal] = Field(None, description="Latest total assets")
    latest_total_equity: Optional[Decimal] = Field(None, description="Latest total equity")
    
    # Key Ratios
    pe_ratio: Optional[float] = Field(None, description="P/E ratio")
    pb_ratio: Optional[float] = Field(None, description="P/B ratio")
    roe: Optional[float] = Field(None, description="Return on equity")
    roa: Optional[float] = Field(None, description="Return on assets")
    debt_to_equity: Optional[float] = Field(None, description="Debt-to-equity ratio")
    
    # Growth Rates (YoY)
    revenue_growth: Optional[float] = Field(None, description="Revenue growth rate")
    earnings_growth: Optional[float] = Field(None, description="Earnings growth rate")
    
    last_updated: datetime = Field(..., description="Last update timestamp")


# Earnings Calendar Schemas
class EarningsCalendarEntry(BaseModel):
    """Earnings calendar entry schema."""
    
    ticker: str = Field(..., description="Stock ticker")
    company_name: str = Field(..., description="Company name")
    earnings_date: date = Field(..., description="Earnings announcement date")
    fiscal_year: int = Field(..., description="Fiscal year")
    fiscal_quarter: int = Field(..., description="Fiscal quarter")
    eps_estimate: Optional[Decimal] = Field(None, description="EPS estimate")
    revenue_estimate: Optional[Decimal] = Field(None, description="Revenue estimate")
    confirmed: bool = Field(False, description="Date confirmed by company")
    market_time: str = Field("AMC", description="Market timing (BMO, AMC, DMT)")
    
    @validator('market_time')
    def validate_market_time(cls, v):
        """Validate market timing."""
        if v not in ['BMO', 'AMC', 'DMT']:  # Before Market Open, After Market Close, During Market Time
            raise ValueError('Market time must be BMO, AMC, or DMT')
        return v


class EarningsCalendarResponse(BaseModel):
    """Earnings calendar response schema."""
    
    date_range: str = Field(..., description="Date range for calendar")
    entries: List[EarningsCalendarEntry] = Field(..., description="Calendar entries")
    total_count: int = Field(..., description="Total number of entries")


# Paginated Financial Responses
class PaginatedFinancialReportsResponse(PaginatedResponse):
    """Paginated financial reports response."""
    
    items: List[FinancialReport]


class PaginatedEarningsResponse(PaginatedResponse):
    """Paginated earnings response."""
    
    items: List[Earnings]


class PaginatedBalanceSheetsResponse(PaginatedResponse):
    """Paginated balance sheets response."""
    
    items: List[BalanceSheet]


class PaginatedIncomeStatementsResponse(PaginatedResponse):
    """Paginated income statements response."""
    
    items: List[IncomeStatement]


class PaginatedCashFlowsResponse(PaginatedResponse):
    """Paginated cash flows response."""
    
    items: List[CashFlow]

File: frontend/src/types/financial.ts
/**
 * Financial data TypeScript interfaces
 */

import { BaseEntity, DateString, PaginatedResponse, Timestamp } from './base';

// Financial Report Types
export type FiscalPeriod = 'Q1' | 'Q2' | 'Q3' | 'Q4' | 'FY' | 'H1' | 'H2';
export type ReportType = 'earnings' | 'balance_sheet' | 'cash_flow' | 'income_statement';

export interface FinancialReport extends BaseEntity {
  ticker: string;
  fiscal_year: number;
  fiscal_period: FiscalPeriod;
  report_type: ReportType;
  announced_at: Timestamp;
  source_url?: string;
  data: Record<string, any>;
}

export interface FinancialReportCreate {
  ticker: string;
  fiscal_year: number;
  fiscal_period: FiscalPeriod;
  report_type: ReportType;
  announced_at: Timestamp;
  source_url?: string;
  data: Record<string, any>;
}

// Earnings Types
export interface Earnings extends BaseEntity {
  ticker: string;
  fiscal_year: number;
  fiscal_quarter: number;
  earnings_date: DateString;
  revenue?: number;
  net_income?: number;
  eps_actual?: number;
  eps_estimate?: number;
  eps_surprise?: number;
  revenue_estimate?: number;
  revenue_surprise?: number;
}

export interface EarningsCreate {
  ticker: string;
  fiscal_year: number;
  fiscal_quarter: number;
  earnings_date: DateString;
  revenue?: number;
  net_income?: number;
  eps_actual?: number;
  eps_estimate?: number;
  eps_surprise?: number;
  revenue_estimate?: number;
  revenue_surprise?: number;
}

// Balance Sheet Types
export interface BalanceSheet extends BaseEntity {
  ticker: string;
  fiscal_year: number;
  fiscal_period: FiscalPeriod;
  report_date: DateString;
  
  // Assets
  total_assets?: number;
  current_assets?: number;
  cash_and_equivalents?: number;
  accounts_receivable?: number;
  inventory?: number;
  property_plant_equipment?: number;
  
  // Liabilities
  total_liabilities?: number;
  current_liabilities?: number;
  accounts_payable?: number;
  short_term_debt?: number;
  long_term_debt?: number;
  
  // Equity
  total_equity?: number;
  retained_earnings?: number;
}

export interface BalanceSheetCreate {
  ticker: string;
  fiscal_year: number;
  fiscal_period: FiscalPeriod;
  report_date: DateString;
  total_assets?: number;
  current_assets?: number;
  cash_and_equivalents?: number;
  accounts_receivable?: number;
  inventory?: number;
  property_plant_equipment?: number;
  total_liabilities?: number;
  current_liabilities?: number;
  accounts_payable?: number;
  short_term_debt?: number;
  long_term_debt?: number;
  total_equity?: number;
  retained_earnings?: number;
}

// Income Statement Types
export interface IncomeStatement extends BaseEntity {
  ticker: string;
  fiscal_year: number;
  fiscal_period: FiscalPeriod;
  report_date: DateString;
  
  // Revenue
  total_revenue?: number;
  cost_of_revenue?: number;
  gross_profit?: number;
  
  // Operating
  operating_expenses?: number;
  operating_income?: number;
  
  // Other
  interest_expense?: number;
  interest_income?: number;
  other_income?: number;
  
  // Income
  income_before_tax?: number;
  income_tax_expense?: number;
  net_income?: number;
  
  // Per Share
  basic_eps?: number;
  diluted_eps?: number;
  weighted_average_shares?: number;
}

export interface IncomeStatementCreate {
  ticker: string;
  fiscal_year: number;
  fiscal_period: FiscalPeriod;
  report_date: DateString;
  total_revenue?: number;
  cost_of_revenue?: number;
  gross_profit?: number;
  operating_expenses?: number;
  operating_income?: number;
  interest_expense?: number;
  interest_income?: number;
  other_income?: number;
  income_before_tax?: number;
  income_tax_expense?: number;
  net_income?: number;
  basic_eps?: number;
  diluted_eps?: number;
  weighted_average_shares?: number;
}

// Cash Flow Types
export interface CashFlow extends BaseEntity {
  ticker: string;
  fiscal_year: number;
  fiscal_period: FiscalPeriod;
  report_date: DateString;
  
  // Operating Activities
  net_income?: number;
  depreciation_amortization?: number;
  changes_working_capital?: number;
  operating_cash_flow?: number;
  
  // Investing Activities
  capital_expenditures?: number;
  investments?: number;
  investing_cash_flow?: number;
  
  // Financing Activities
  debt_issuance?: number;
  debt_repayment?: number;
  dividends_paid?: number;
  share_repurchases?: number;
  financing_cash_flow?: number;
  
  // Net Change
  net_change_cash?: number;
  free_cash_flow?: number;
}

export interface CashFlowCreate {
  ticker: string;
  fiscal_year: number;
  fiscal_period: FiscalPeriod;
  report_date: DateString;
  net_income?: number;
  depreciation_amortization?: number;
  changes_working_capital?: number;
  operating_cash_flow?: number;
  capital_expenditures?: number;
  investments?: number;
  investing_cash_flow?: number;
  debt_issuance?: number;
  debt_repayment?: number;
  dividends_paid?: number;
  share_repurchases?: number;
  financing_cash_flow?: number;
  net_change_cash?: number;
  free_cash_flow?: number;
}

// Financial Ratios Types
export interface FinancialRatios extends BaseEntity {
  ticker: string;
  fiscal_year: number;
  fiscal_period: FiscalPeriod;
  calculation_date: DateString;
  
  // Profitability Ratios
  gross_margin?: number;
  operating_margin?: number;
  net_margin?: number;
  return_on_assets?: number;
  return_on_equity?: number;
  
  // Liquidity Ratios
  current_ratio?: number;
  quick_ratio?: number;
  cash_ratio?: number;
  
  // Leverage Ratios
  debt_to_equity?: number;
  debt_to_assets?: number;
  interest_coverage?: number;
  
  // Efficiency Ratios
  asset_turnover?: number;
  inventory_turnover?: number;
  receivables_turnover?: number;
}

export interface FinancialRatiosCreate {
  ticker: string;
  fiscal_year: number;
  fiscal_period: FiscalPeriod;
  calculation_date: DateString;
  gross_margin?: number;
  operating_margin?: number;
  net_margin?: number;
  return_on_assets?: number;
  return_on_equity?: number;
  current_ratio?: number;
  quick_ratio?: number;
  cash_ratio?: number;
  debt_to_equity?: number;
  debt_to_assets?: number;
  interest_coverage?: number;
  asset_turnover?: number;
  inventory_turnover?: number;
  receivables_turnover?: number;
}

// Financial Summary Types
export interface FinancialSummary {
  ticker: string;
  company_name: string;
  latest_fiscal_year: number;
  latest_fiscal_period: string;
  
  // Latest Financial Data
  latest_revenue?: number;
  latest_net_income?: number;
  latest_eps?: number;
  latest_total_assets?: number;
  latest_total_equity?: number;
  
  // Key Ratios
  pe_ratio?: number;
  pb_ratio?: number;
  roe?: number;
  roa?: number;
  debt_to_equity?: number;
  
  // Growth Rates (YoY)
  revenue_growth?: number;
  earnings_growth?: number;
  
  last_updated: Timestamp;
}

// Earnings Calendar Types
export interface EarningsCalendarEntry {
  ticker: string;
  company_name: string;
  earnings_date: DateString;
  fiscal_year: number;
  fiscal_quarter: number;
  eps_estimate?: number;
  revenue_estimate?: number;
  confirmed: boolean;
  market_time: 'BMO' | 'AMC' | 'DMT'; // Before Market Open, After Market Close, During Market Time
}

export interface EarningsCalendarResponse {
  date_range: string;
  entries: EarningsCalendarEntry[];
  total_count: number;
}

// Financial Analysis Types
export interface FinancialAnalysis {
  ticker: string;
  analysis_date: Timestamp;
  
  // Trend Analysis
  revenue_trend: {
    direction: 'up' | 'down' | 'stable';
    growth_rate: number;
    consistency_score: number;
  };
  
  profitability_trend: {
    direction: 'up' | 'down' | 'stable';
    margin_improvement: number;
    efficiency_score: number;
  };
  
  // Financial Health
  liquidity_score: number; // 0-100
  solvency_score: number; // 0-100
  efficiency_score: number; // 0-100
  overall_health_score: number; // 0-100
  
  // Peer Comparison
  peer_comparison: {
    revenue_percentile: number;
    profitability_percentile: number;
    efficiency_percentile: number;
    valuation_percentile: number;
  };
  
  // Key Insights
  strengths: string[];
  weaknesses: string[];
  opportunities: string[];
  risks: string[];
}

// Financial Chart Data Types
export interface FinancialChartData {
  ticker: string;
  chart_type: 'revenue' | 'earnings' | 'margins' | 'ratios' | 'cash_flow';
  time_series: {
    period: string;
    value: number;
    label: string;
  }[];
  comparison_data?: {
    ticker: string;
    company_name: string;
    time_series: {
      period: string;
      value: number;
    }[];
  }[];
}

// Financial Metrics Comparison Types
export interface FinancialComparison {
  tickers: string[];
  comparison_type: 'peer_analysis' | 'sector_analysis' | 'historical_analysis';
  metrics: {
    [ticker: string]: {
      revenue: number;
      net_income: number;
      eps: number;
      roe: number;
      roa: number;
      debt_to_equity: number;
      current_ratio: number;
      pe_ratio: number;
      pb_ratio: number;
    };
  };
  rankings: {
    metric: string;
    rankings: {
      ticker: string;
      value: number;
      rank: number;
    }[];
  }[];
}

// Financial State Types (for Redux/Context)
export interface FinancialState {
  // Current financial data
  currentFinancials: {
    [ticker: string]: {
      earnings: Earnings[];
      balanceSheet: BalanceSheet[];
      incomeStatement: IncomeStatement[];
      cashFlow: CashFlow[];
      ratios: FinancialRatios[];
      summary: FinancialSummary;
    };
  };
  
  // Earnings calendar
  earningsCalendar: EarningsCalendarEntry[];
  earningsCalendarDateRange: string;
  
  // Financial analysis
  financialAnalysis: {
    [ticker: string]: FinancialAnalysis;
  };
  
  // Chart data
  chartData: {
    [ticker: string]: {
      [chartType: string]: FinancialChartData;
    };
  };
  
  // Loading states
  isLoadingFinancials: boolean;
  isLoadingEarningsCalendar: boolean;
  isLoadingAnalysis: boolean;
  isLoadingChartData: boolean;
  
  // Error states
  financialsError?: string;
  earningsCalendarError?: string;
  analysisError?: string;
  chartDataError?: string;
  
  // Last updated timestamps
  lastUpdated: {
    financials?: Timestamp;
    earningsCalendar?: Timestamp;
    analysis?: Timestamp;
    chartData?: Timestamp;
  };
}

// Financial Export Types
export interface FinancialDataExport {
  ticker: string;
  export_type: 'earnings' | 'balance_sheet' | 'income_statement' | 'cash_flow' | 'all';
  format: 'csv' | 'excel' | 'json';
  date_range: {
    start_year: number;
    end_year: number;
  };
  include_estimates: boolean;
  include_ratios: boolean;
}

// Paginated Financial Responses
export type PaginatedFinancialReportsResponse = PaginatedResponse<FinancialReport>;
export type PaginatedEarningsResponse = PaginatedResponse<Earnings>;
export type PaginatedBalanceSheetsResponse = PaginatedResponse<BalanceSheet>;
export type PaginatedIncomeStatementsResponse = PaginatedResponse<IncomeStatement>;
export type PaginatedCashFlowsResponse = PaginatedResponse<CashFlow>;